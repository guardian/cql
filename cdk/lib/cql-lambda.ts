import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export class CqlLambda extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		new GuLambdaFunction(this, 'cql-lambda', {
			app: 'cql-lambda',
			handler: 'cql.Handler::handleRequest',
			functionName: `cql-lambda-${this.stage}`,
      runtime: Runtime.JAVA_11,
      fileName: "cql-lambda.jar"
		});
	}
}
