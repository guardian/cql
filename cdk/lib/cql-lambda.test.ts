import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { CqlLambda } from "./cql-lambda";

describe("The CqlLambda stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new CqlLambda(app, "CqlLambda", { stack: "cql", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
