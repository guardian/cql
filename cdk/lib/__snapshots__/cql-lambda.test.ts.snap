// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CqlLambda stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuApiLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "cqllambdaqueryEndpoint221F8116": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "cqllambdaqueryC5900F49",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "cqllambdaqueryDeploymentStageprod1835B037",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "cqllambdaF47365BA": {
      "DependsOn": [
        "cqllambdaServiceRoleDefaultPolicyF7B612C2",
        "cqllambdaServiceRoleACC97F7F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "cql/TEST/cql-lambda/cql-lambda.jar",
        },
        "Environment": {
          "Variables": {
            "APP": "cql-lambda",
            "STACK": "cql",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "cql-lambda-TEST",
        "Handler": "cql.Handler::handleRequest",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "cqllambdaServiceRoleACC97F7F",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "cql-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cql",
          },
          {
            "Key": "Stack",
            "Value": "cql",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "cqllambdaServiceRoleACC97F7F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "cql-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cql",
          },
          {
            "Key": "Stack",
            "Value": "cql",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "cqllambdaServiceRoleDefaultPolicyF7B612C2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/cql/TEST/cql-lambda/cql-lambda.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/cql/cql-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/cql/cql-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "cqllambdaServiceRoleDefaultPolicyF7B612C2",
        "Roles": [
          {
            "Ref": "cqllambdaServiceRoleACC97F7F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "cqllambdaqueryANYA3DB3FA1": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "cqllambdaF47365BA",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "cqllambdaqueryC5900F49",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "cqllambdaqueryC5900F49",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "cqllambdaqueryANYApiPermissionCqlLambdacqllambdaquery8BCDECD0ANY49E8C1C5": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "cqllambdaF47365BA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "cqllambdaqueryC5900F49",
              },
              "/",
              {
                "Ref": "cqllambdaqueryDeploymentStageprod1835B037",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "cqllambdaqueryANYApiPermissionTestCqlLambdacqllambdaquery8BCDECD0ANYAC57CB1D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "cqllambdaF47365BA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "cqllambdaqueryC5900F49",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "cqllambdaqueryAccountFAB0BCC8": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "cqllambdaqueryC5900F49",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "cqllambdaqueryCloudWatchRole894EFE1B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "cqllambdaqueryC5900F49": {
      "Properties": {
        "Description": "Get a CAPI query string from a CQL query",
        "Name": "cql-TEST-query",
        "Tags": [
          {
            "Key": "App",
            "Value": "cql-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cql",
          },
          {
            "Key": "Stack",
            "Value": "cql",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "cqllambdaqueryCloudWatchRole894EFE1B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "cql-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cql",
          },
          {
            "Key": "Stack",
            "Value": "cql",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "cqllambdaqueryDeployment449DA6F564134b68fde643afce1003e622b9c5c4": {
      "DependsOn": [
        "cqllambdaqueryproxyANY3559EDCF",
        "cqllambdaqueryproxy5B38C941",
        "cqllambdaqueryANYA3DB3FA1",
      ],
      "Properties": {
        "Description": "Get a CAPI query string from a CQL query",
        "RestApiId": {
          "Ref": "cqllambdaqueryC5900F49",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "cqllambdaqueryDeploymentStageprod1835B037": {
      "DependsOn": [
        "cqllambdaqueryAccountFAB0BCC8",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "cqllambdaqueryDeployment449DA6F564134b68fde643afce1003e622b9c5c4",
        },
        "RestApiId": {
          "Ref": "cqllambdaqueryC5900F49",
        },
        "StageName": "prod",
        "Tags": [
          {
            "Key": "App",
            "Value": "cql-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cql",
          },
          {
            "Key": "Stack",
            "Value": "cql",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "cqllambdaqueryproxy5B38C941": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "cqllambdaqueryC5900F49",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "cqllambdaqueryC5900F49",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "cqllambdaqueryproxyANY3559EDCF": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "cqllambdaF47365BA",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "cqllambdaqueryproxy5B38C941",
        },
        "RestApiId": {
          "Ref": "cqllambdaqueryC5900F49",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "cqllambdaqueryproxyANYApiPermissionCqlLambdacqllambdaquery8BCDECD0ANYproxy82F11349": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "cqllambdaF47365BA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "cqllambdaqueryC5900F49",
              },
              "/",
              {
                "Ref": "cqllambdaqueryDeploymentStageprod1835B037",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "cqllambdaqueryproxyANYApiPermissionTestCqlLambdacqllambdaquery8BCDECD0ANYproxy0B6F6F7E": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "cqllambdaF47365BA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "cqllambdaqueryC5900F49",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
